File: scala/src/main/scala/nonamedb/storage/engines/memory/MemoryEngine.scala
package nonamedb.storage.engines.memory

import akka.Done
import akka.actor.{ActorRef, ActorSystem}
import akka.pattern.ask
import akka.util.Timeout
import io.opentelemetry.api.trace.{Span, Tracer}
import nonamedb.storage.engines.memory.EngineActor.{Get, Put}
import nonamedb.storage.{Engine, Key, Value}

import scala.concurrent.Future

class MemoryEngine(tracer: Tracer)(implicit system: ActorSystem, timeout: Timeout) extends Engine {
  private val store: ActorRef = system.actorOf(EngineActor.props(tracer))

  override def get(key: Key)(implicit span: Span): Future[Option[Value]] =
    (store ? Get(key, span)).mapTo[Option[Value]]

  override def put(key: Key, value: Value)(implicit span: Span): Future[Done] =
    (store ? Put(key, value, span)).mapTo[Done]
}

object MemoryEngine {
  def apply(tracer: Tracer)(implicit system: ActorSystem, timeout: Timeout): MemoryEngine =
    new MemoryEngine(tracer)
}

File: python/nonamedb/storage/engines/memory/MemoryEngine.py
from nonamedb.storage.Engine import Engine
from nonamedb.storage.engines.memory.EngineActor import EngineActor, Get, Put


class MemoryEngine(Engine):
    def __init__(self, actor_system, request_timeout):
        self.__system = actor_system
        self.__timeout = request_timeout
        self.__store = self.__system.createActor(EngineActor)

    def get(self, key):
        result = self.__system.ask(self.__store, Get(key), self.__timeout)
        return result

    def put(self, key, value):
        result = self.__system.ask(self.__store, Put(key, value), self.__timeout)
        return result
